cloudnew/ec2.go

```diff
@@ -0,0 +1,518 @@
+		resp, err = m.client.DescribeInstances(&ec2.DescribeInstancesInput{
+			InstanceIds: []*string{instance.InstanceId},
+		})
+		if err != nil {
+			if i == max {
+				err = errors.Wrap(err, "error querying for instance info and retries exhausted")
+				grip.Error(message.Fields{
+					"error":    err,
+					"instance": *instance.InstanceId,
+				})
+				return nil, err
+			}
+			time.Sleep(30 * time.Second)
+			continue
+		}
+		break
+	}
+
+	grip.Debug(message.Fields{
+		"msg":      "describe instances returned data",
+		"host_id":  h.Id,
+		"response": resp,
+	})
+
+	if len(resp.Reservations) < 1 {
+		return nil, errors.New("describe instances response has no reservations")
+	}
+
+	instance = resp.Reservations[0].Instances[0]
+	h.Id = *instance.InstanceId
+	resources := []*string{instance.InstanceId}
+	for _, vol := range instance.BlockDeviceMappings {
+		if *vol.DeviceName == "" {
+			continue
+		}
+
+		resources = append(resources, vol.Ebs.VolumeId)
+	}
+	return resources, nil
+}
+
+func (m *ec2Manager) spawnSpotHost(h *host.Host, ec2Settings *EC2ProviderSettings, blockDevices []*ec2.BlockDeviceMapping) ([]*string, error) {
+	spotRequest := &ec2.RequestSpotInstancesInput{
+		SpotPrice:     makeStringPtr(fmt.Sprintf("%v", ec2Settings.BidPrice)),
+		InstanceCount: makeInt64Ptr(1),
+		LaunchSpecification: &ec2.RequestSpotLaunchSpecification{
+			ImageId:             makeStringPtr(ec2Settings.AMI),
+			KeyName:             makeStringPtr(ec2Settings.KeyName),
+			InstanceType:        makeStringPtr(ec2Settings.InstanceType),
+			BlockDeviceMappings: blockDevices,
+		},
+	}
+
+	if ec2Settings.IsVpc {
+		spotRequest.LaunchSpecification.SecurityGroupIds = []*string{&ec2Settings.SecurityGroup}
+		spotRequest.LaunchSpecification.SubnetId = &ec2Settings.SubnetId
+	} else {
+		spotRequest.LaunchSpecification.SecurityGroups = []*string{&ec2Settings.SecurityGroup}
+	}
+
+	spotResp, err := m.client.RequestSpotInstances(spotRequest)
+	if err != nil {
+		grip.Error(errors.Wrapf(h.Remove(), "error removing intent host %s", h.Id))
+		return nil, errors.Wrap(err, "RequestSpotInstances API call returned an error")
+	}
+
+	spotReqRes := spotResp.SpotInstanceRequests[0]
+	if *spotReqRes.State != cloud.SpotStatusOpen && *spotReqRes.State != cloud.SpotStatusActive {
+		err = errors.Errorf("Spot request %s was found in state %s on intent host %s",
+			*spotReqRes.SpotInstanceRequestId, *spotReqRes.State, h.Id)
+		grip.Error(err)
+		return nil, err
+	}
+
+	h.Id = *spotReqRes.SpotInstanceRequestId
+	resources := []*string{spotReqRes.SpotInstanceRequestId}
+	return resources, nil
+}
+
+// SpawnHost spawns a new host.
+func (m *ec2Manager) SpawnHost(h *host.Host) (*host.Host, error) {
+	if h.Distro.Provider != evergreen.ProviderNameEc2OnDemand && h.Distro.Provider != evergreen.ProviderNameEc2Spot {
+		return nil, errors.Errorf("Can't spawn instance of %s for distro %s: provider is %s",
+			evergreen.ProviderNameEc2OnDemand, h.Distro.Id, h.Distro.Provider)
```

Comment:
maybe it makes sense here, and for other errors from this function, to do the logging here (rather than at the call site) just so we can put more structured information in a place that its useful. 

the function is of multiple minds, it seems, about where to log in error cases.

Reply:
fair, but I think we loose some fidelity by putting all of the metadata into the strings.